src/core/distributions/events/MarkovEventsDistribution.h

```diff
@@ -0,0 +1,374 @@
+    }
+    
+    this->value = simulate();
+}
+
+template <class valueType>
+RevBayesCore::MarkovEventsDistribution<valueType>::~MarkovEventsDistribution<valueType>( void )
+{
+}
+
+
+
+template <class valueType>
+RevBayesCore::MarkovEventsDistribution<valueType>* RevBayesCore::MarkovEventsDistribution<valueType>::clone( void ) const
+{
+    return new MarkovEventsDistribution<valueType>( *this );
+}
+
+template <class valueType>
+double RevBayesCore::MarkovEventsDistribution<valueType>::computeLnProbability( void )
+{
+	// make sure the times and events are one-to-one
+	const std::set<double>& the_event_times = event_times->getValue().getEventTimes();
+	const std::map<double, valueType>& the_values = this->value->getEvents();
+
+	// make sure the number of events are the same
+	if ( the_event_times.size() != the_values.size() )
+	{
+		throw RbException("The number of event times and the number of events do not match up!");
+	}
+
+	// make sure the times match
+	for( std::set<double>::const_iterator it = the_event_times.begin(); it != the_event_times.end(); ++it)
+	{
+		// try to find the event with that time
+		typename std::map<double, valueType>::const_iterator ev = the_values.find( *it );
+
+		// if not found, throw error
+		if ( it == the_event_times.end() ) {
+			throw RbException("The times do not match up with event times");
+		}
+	}
+
+	// compute the probability
+	double ln_prob = 0.0;
+
+	for( typename std::map<double, valueType>::const_iterator it = the_values.begin(); it != the_values.end(); ++it)
+	{
+		// set the value for the base distribution
+		base_distribution->setValue( Cloner<valueType, IsDerivedFrom<valueType, Cloneable>::Is >::createClone( it->second ) );
+
+		// compute the probability
+		ln_prob += base_distribution->computeLnProbability();
+	}
+
+    return ln_prob;
+}
+
+template <class valueType>
+RevBayesCore::OrderedEvents<valueType>* RevBayesCore::MarkovEventsDistribution<valueType>::simulate()
+{
+	// get the number of events from the times
+	const std::set<double>& the_times  = event_times->getValue().getEventTimes();
+	size_t                  num_events = event_times->getValue().size();
+
+	// create the new container
+	OrderedEvents<valueType>* sim = new OrderedEvents<valueType>();
+
+	// for each event, simulate a value
+	for(std::set<double>::const_iterator it = the_times.begin(); it != the_times.end(); ++it)
+	{
+		// have the base distribution simulate a new value
+		base_distribution->redrawValue();
+
+		// copy the simulated value
+		valueType new_event = base_distribution->getValue();
+
+		// add the new event
+		sim->addEvent(*it, new_event);
+	}
+
+	// return the simulation
+	return sim;
+}
+
+
+template <class valueType>
+void RevBayesCore::MarkovEventsDistribution<valueType>::redrawValue( void )
+{
+    delete this->value;
+    this->value = simulate();
+
+    // tell any rep children to update
+	const std::vector<DagNode*>& children = this->dag_node->getChildren();
```

Comment:
This might be more readable as 
`for(auto& child: this->dag_node->getChildren())
{
   child.touch()
}
`

Reply:
See above comment about removing this files from the merge (they're experimental).